### PR TITLE
Add additional targets

### PR DESCRIPTION
**Pull request by ikstream**
In this PR two more build targets are added:
* ``` make flash ``` starts a clean build and flashes to detected ttyUSB
* ``` make monitor ``` opens a minicom serial on detected ttyUSB

In addition the base platformio structure is added to ensure, that the build targets finds everything it expects.    
Furthermore a gitignore file is added to keep build files out of this repository

### CODE REVIEW

File: platformio.ini [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+;   Library options: dependencies, extra library storages
+;   Advanced options: extra scripting
+;
+; Please visit documentation for the other options and examples
+; https://docs.platformio.org/page/projectconf.html
+[env]
+platform = espressif32
+board = pico32
+framework = espidf
+upload_port = /dev/ttyUSB*
```
Comment (nopeslide) [inline platformio.ini]: why hardcode the upload_port if it can be specified by an pio run argument?

File: docker/Makefile [hunk 2]
```diff
@@ -45,6 +48,29 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
...
+.PHONY: flash
+flash:
+	docker run --rm -ti \
+		--device /dev/${DEV} \
+		--volume "$(shell pwd)/..":/mnt:rw \
+		-w /mnt \
+		--name flash \
+		lifesensor/platformio:$(PLATFORMIO_VERSION) \
+		/bin/sh -c \
+		"pio run -t clean; pio run -t upload; /bin/bash"
```
Comment (nopeslide) [inline docker/Makefile]: why start bash?
further, i would remove the clean job.
let multiple jobs be handled by the makefile

File: docker/Makefile [hunk 3]
```diff
@@ -45,6 +48,29 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
 	| tee -a $@.part
 	mv $@.part $@
 
+.PHONY: flash
+flash:
```
Comment (nopeslide) [inline docker/Makefile]: pio actions should be prefixed with `pio-`
an maybe add an `pio-upload` alias?

File: docker/Makefile [hunk 4]
```diff
@@ -45,6 +48,29 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
...
+		--device /dev/${DEV} \
+		--volume "$(shell pwd)/..":/mnt:rw \
+		-w /mnt \
+		--name flash \
+		lifesensor/platformio:$(PLATFORMIO_VERSION) \
+		/bin/sh -c \
+		"pio run -t clean; pio run -t upload; /bin/bash"
+
+.PHONY: monitor
+monitor:
```
Comment (nopeslide) [inline docker/Makefile]: `pio-monitor`

File: docker/Makefile [hunk 5]
```diff
@@ -23,6 +24,8 @@ help:
 	@echo "make platformio        | starts bash in platformio container"
 	@echo "make qemu              | starts bash in qemu container"
 	@echo "make vscode            | starts bash&vscode in vscode container"
+	@echo "make flash             | starts clean build and upload proccess"
```
Comment (nopeslide) [inline docker/Makefile]: `pio-flash` and maybe `pio-upload`?

File: docker/Makefile [hunk 6]
```diff
@@ -23,6 +24,8 @@ help:
...
+	@echo "make monitor           | starts minicom serial console"
```
Comment (nopeslide) [inline docker/Makefile]: `pio-monitor`

File: docker/Makefile [hunk 7]
```diff
@@ -45,6 +48,29 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
...
+.PHONY: monitor
+monitor:
+	docker run --rm -ti \
+		--device /dev/${DEV} \
+		--volume "$(shell pwd)/..":/mnt:rw \
+		--name monitor \
+		lifesensor/platformio:$(PLATFORMIO_VERSION) \
+		/bin/sh -c \
+		"python3 \
+		/home/developer/.platformio/packages/framework-espidf/tools/idf_monitor.py \
```
Comment (nopeslide) [inline docker/Makefile:72]: specify the port

File: test/README [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+This directory is intended for PIO Unit Testing and project tests.
+
+Unit Testing is a software testing method by which individual units of
+source code, sets of one or more MCU program modules together with associated
+control data, usage procedures, and operating procedures, are tested to
+determine whether they are fit for use. Unit testing finds problems early
+in the development cycle.
+
+More information about PIO Unit Testing:
+- https://docs.platformio.org/page/plus/unit-testing.html
```
Comment (nopeslide) [inline test/README]: add file suffix `.md`

File: src/sdkconfig [hunk 9]
```diff
@@ -0,0 +1,962 @@
+#
```
Comment (nopeslide) [inline src/sdkconfig:1]: is this file necessary since it gets autogenerated?

File: src/sdkconfig.h [hunk 10]
```diff
@@ -0,0 +1,371 @@
+/*
```
Comment (nopeslide) [inline src/sdkconfig.h:1]: is this file necessary since its autogenerated?

File: lib/README [hunk 11]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (nopeslide) [inline lib/README]: add `.md` suffix

File: include/README [hunk 12]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (nopeslide) [inline include/README]: add `.md` suffix

**Review by nopeslide — CHANGES_REQUESTED** [review]

File: docker/Makefile [hunk 13]
```diff
@@ -1,5 +1,6 @@
 PLATFORMIO_VERSION?=4.1.0
 QEMU_COMMIT?=esp-develop
+DEV=$(shell dmesg | awk '/cp210x/ && /ttyUSB/ { print $$NF }')
```
Comment (nopeslide) [inline docker/Makefile]: change to `DEV?=`

Comment (nopeslide) [reply docker/Makefile]: is sth like `readlink -f /dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller*` maybe better?

File: src/sdkconfig (hunk 9 above)
Comment (ikstream) [reply src/sdkconfig:1]: It needs to be generated at the first run and wont be regenerated later

File: src/sdkconfig.h (hunk 10 above)
Comment (ikstream) [reply src/sdkconfig.h:1]: see comment above

Comment (nopeslide) [reply src/sdkconfig.h:1]: this file is generated by `idf.py menuconfig` or similar, but platformio doesn't seem to support it and recommends to modify `sdkconfig.h` directly.
https://github.com/platformio/platform-espressif32/issues/20
https://community.platformio.org/t/esp32-idf-menuconfig-platofrmio-dragging-their-feet/6268
should we also install idf in the container and use their menuconfig or wait this out?

File: platformio.ini (hunk 1 above)
Comment (ikstream) [reply platformio.ini]: If several ttyUSB (esp. ESP32 of the same type) devices are connected the individual user can select which ttyUSB to use. So i would prefer this version over an argument in a dockerfile

File: docker/Makefile (hunk 2 above)
Comment (ikstream) [reply docker/Makefile]: I agree

File: src/sdkconfig.h (hunk 10 above)
Comment (ikstream) [reply src/sdkconfig.h:1]: I would say we wait this out as pio allows us to modify the ```sdkconfig.h``` file there is no need to increase complexity

File: docker/Makefile (hunk 13 above)
Comment (ikstream) [reply docker/Makefile]: I would say the check for cp210x is more generic, but i would need to test some more different devices to be sure

Comment (ikstream) [reply docker/Makefile]: Ok, this version wont work if a ttyUSB was connected and removed later. It will still show up in dmesg, but isn't present in /dev/ anymore.

**Review by nopeslide — APPROVED** [review]

**Review by nopeslide — CHANGES_REQUESTED** [review]
accidentally accepted the PR

File: src/sdkconfig.h (hunk 10 above)
Comment (nopeslide) [reply src/sdkconfig.h:1]: ok

File: platformio.ini (hunk 1 above)
Comment (nopeslide) [reply platformio.ini]: the device can be selected via make like `make pio-flash DEV=/dev/ttyUSB1` without touching any files. the Makefile could be extended by a device check if the fallback doesn't exist, or even choose a device

File: docker/Makefile (hunk 13 above)
Comment (ikstream) [reply docker/Makefile]: how about adding DEV variable to make to let user specify (esp. with several ttyUSB devices) which devices to use?

Comment (nopeslide) [reply docker/Makefile]: I don't think the reconnect issue would lead to any major problems, at least not in the current state.

Comment (nopeslide) [reply docker/Makefile]: your wish for a generic solution is also possible via globbing `readlink -f /dev/serial/by-id/*CP210*`. To be honest, I just dislike the idea of "parsing" `dmesg`

Comment (nopeslide) [reply docker/Makefile]: > how about adding DEV variable to make to let user specify (esp. with several ttyUSB devices) which devices to use?

you already achieved this by adding the DEV variable, see my comment on the `platformio.ini`. But you have to make it a fallback via `?=`

File: docker/Makefile [hunk 14]
```diff
@@ -48,28 +49,39 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
 	| tee -a $@.part
 	mv $@.part $@
 
-.PHONY: flash
-flash:
+.PHONY: pio-flash pio-upload
+pio-flash pio-upload:
```
Comment (nopeslide) [inline docker/Makefile]: the container dependency and check if docker is installed is missing
`pio-flash pio-upload: | docker-platformio check-docker`

File: docker/Makefile [hunk 15]
```diff
@@ -48,28 +49,39 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
...
 		--name flash \
 		lifesensor/platformio:$(PLATFORMIO_VERSION) \
 		/bin/sh -c \
-		"pio run -t clean; pio run -t upload; /bin/bash"
+		"pio run -t upload"
 
-.PHONY: monitor
-monitor:
+.PHONY: pio-monitor
+pio-monitor:
```
Comment (nopeslide) [inline docker/Makefile]: the container dependency and check if docker is installed is missing
`pio-flash pio-upload: | docker-platformio check-docker`

File: docker/Makefile [hunk 16]
```diff
@@ -48,28 +49,39 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
...
 		--name monitor \
 		lifesensor/platformio:$(PLATFORMIO_VERSION) \
 		/bin/sh -c \
 		"python3 \
 		/home/developer/.platformio/packages/framework-espidf/tools/idf_monitor.py \
-		/mnt/.pio/build/pico32/firmware.elf"
+		/mnt/.pio/build/pico32/firmware.elf --port ${DEV}"
+
+.PHONY: pio-clean
+pio-clean:
```
Comment (nopeslide) [inline docker/Makefile]: the container dependency and check if docker is installed is missing
`pio-flash pio-upload: | docker-platformio check-docker`

File: docker/Makefile [hunk 17]
```diff
@@ -48,28 +49,39 @@ platformio/Dockerfile.log: platformio/Dockerfile | check-docker
...
 	docker run --rm -ti \
-		--device /dev/${DEV} \
+		--device "${DEV}" \
 		--volume "$(shell pwd)/..":/mnt:rw \
 		-w /mnt \
 		--name flash \
 		lifesensor/platformio:$(PLATFORMIO_VERSION) \
 		/bin/sh -c \
-		"pio run -t clean; pio run -t upload; /bin/bash"
+		"pio run -t upload"
```
Comment (nopeslide) [inline docker/Makefile]: would add `--upload-port $(DEV)` if we manage the port in the makefile

**Review by nopeslide — CHANGES_REQUESTED** [review]

File: platformio.ini (hunk 1 above)
Comment (nopeslide) [reply platformio.ini]: will we keep the `upload_port`? Port is now managed by the makefile

**Review by nopeslide — APPROVED** [review]
